### PR TITLE
Feature 14 scheduled crawls

### PR DESCRIPTION
**Pull request by Jdallos**

### CODE REVIEW

File: src/functions/index.ts [hunk 1]
```diff
@@ -1,4 +1,9 @@
...
 export { default as listTimedCurations } from './list-timed-curations';
 export { default as updateTimedCuration } from './update-timed-curation';
-export { default as deleteTimedCuration } from './delete-timed-curation';
\ No newline at end of file
+export { default as deleteTimedCuration } from './delete-timed-curation';
+export { default as createScheduledCrawl } from './scheduled-crawls/create-scheduled-crawl';
+export { default as listScheduledCrawls } from './scheduled-crawls/list-scheduled-crawls';
+export { default as updateScheduledCrawl } from './scheduled-crawls/update-scheduled-crawl';
+export { default as deleteScheduledCrawl } from './scheduled-crawls/delete-scheduled-crawl';
+// TODO: Is this missing find-scheduled-crawl, find-timed-curation?
```
Comment (anttileppa) [inline src/functions/index.ts:9]: Resolve this TODO (by adding missing functions)

File: src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  });
+
+  const activeCrawl = await elastic.checkIfActiveCrawl();
+
+  if (activeCrawl.id) {
+    console.info("Crawl already running, current scheduled crawls delayed", activeCrawl);
+  } else {
+    const scheduledCrawls = await scheduledCrawlService.listScheduledCrawls();
+
+    let crawlDetailsMap: { [key: string]: GetCrawlerCrawlRequestResponse };
```
Comment (anttileppa) [inline src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:26]: this should be const

Comment (anttileppa) [inline src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:26]: initialize the map

File: src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      let crawlDetails: GetCrawlerCrawlRequestResponse = crawlDetailsMap[previousCrawlId];
+      if (crawlDetails === undefined) {
+        crawlDetails = previousCrawlId ? await elastic.findCrawlDetails({ id: previousCrawlId }) : null;
+        crawlDetailsMap[previousCrawlId] = crawlDetails;
+      }
+
+      if (!crawlDetails) {
+        return true;
+      } else {
+        const { completed_at } = crawlDetailsMap[previousCrawlId];
```
Comment (anttileppa) [inline src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:40]:
```suggestion
        const { completed_at } = crawlDetails;
```

File: src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    try {
+    const crawlResponse = await elastic.createCrawlRequest(crawlOptions);
+
+    activeCrawls.forEach(activeCrawl => {
+      scheduledCrawlService.updateScheduledCrawl({
+        ...activeCrawl,
+        previousCrawlId: crawlResponse
+      });
+    });
```
Comment (anttileppa) [inline src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:63]: Fix indentation

File: src/libs/date-utils.ts [hunk 5]
```diff
@@ -26,4 +26,20 @@ export const parseDate = (date?: string): Date | undefined => {
...
+ * @param lastCrawl date of last crawl
+ * @returns difference in minutes
+ */
+ export const calculateMinutesPassed = (lastCrawl: string ) => {
+  const lastCrawlDate = parseDate(lastCrawl);
+  const now = new Date();
+
+  const difference = now.getTime() - lastCrawlDate?.getTime();
+  // TODO: +180 accounting for GMT+3, will need improving
+  return Math.round(difference / 60000) + 180;
```
Comment (anttileppa) [inline src/libs/date-utils.ts:44]: This project uses luxon which supports timezone handling by default. Refactor to use it

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/functions/index.ts (hunk 1 above)
Comment (Jdallos) [reply src/functions/index.ts:9]: All missing functions added

File: src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts (hunk 2 above)
Comment (Jdallos) [reply src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:26]: Done

Comment (Jdallos) [reply src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:26]: Done

File: src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts (hunk 3 above)
Comment (Jdallos) [reply src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:40]: Done

File: src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts (hunk 4 above)
Comment (Jdallos) [reply src/functions/scheduled-crawls/trigger-scheduled-crawl/handler.ts:63]: whoops, done

File: src/libs/date-utils.ts (hunk 5 above)
Comment (Jdallos) [reply src/libs/date-utils.ts:44]: I've confused myself thinking about this.
If the elastic and lambda are hosted in the same location no time zone adjustment is needed?
If they are not in the same place I need to find out where they are and adjust accordingly...

**Review by anttileppa — APPROVED** [review]

**Pull request merged**